### DISCUSSION TITLE
Mac jamulus client using -6 option doesn't list directory servers in any genres

### DISCUSSION THREAD

**Discussion by rdica** [jamulussoftware/jamulus#2224] [Bug Investigation]
In testing IPv6 connectivity from a Mac client, I found I could direct connect to an IPv6 enabled server but I could not see any servers listed in the Connect dialog for any genre, nor could I direct connect to a server via IPv4.
It appears connectivity is IPv6-only when using `-6` option.
Running jamulus without `-6` option re-enables IPv4 connectivity for directory server listings for each genre, and direct IPv4 connectivity to servers.

I'm starting jamulus from iTerm2 using this commandline:
`/Applications/Jamulus.app/Contents/MacOS/Jamulus -6`

Under Linux, using `-6` option does not disable IPv4 connectivity.

**Comment by pljones** [comment]
Directories do not support IPv6 addressing.  Try running with `--showallservers` - I'd expect you to get the server list returned (maybe).  However, none of the entries in the server list supplies an IPv6 address.  IPv6 support remains experimental.

**Reply by rdica** [reply]
BTW, `--showallservers` is not shown when `--help` is invoked.
Using that option had no effect, I still could not connect to a server using IPv4, nor did I receive any server listings from directory servers.

I understand that IPv6 support is experimental, hence my testing. Seems to work as expected under Linux, can see and reach servers on both protocols when using `-6`. On Macs the option seems to disable IPv4 altogether, thats not expected or desired behaviour.

Thanks for responding, I wanted to report what I'd found.

**Reply by pljones** [reply]
> BTW, --showallservers is not shown when --help is invoked.

That's intentional.  It's for debugging situations like this, only.

> Using that option had no effect, I still could not connect to a server using IPv4, nor did I receive any server listings from directory servers.

Hm.  OK.  Interesting.

Unfortunately, my ISP doesn't support IPv6, so I can't do any testing.

@softins can you confirm whether the fallback from IPv6 to IPv4 caters for _querying_ a directory?  The directory should only be listening on IPv4, so there should be no IPv6 route.  (I _know_ none of mine have an IPv6 route.)

**Reply by softins** [reply]
OK, I'll do some experiments. I have IPv6 connectivity here.

**Reply by pljones** [reply]
I will note that my Windows and Linux boxes both _have_ IPv6 addresses and, when I run the client with `-6`, I can see _all_ directories - including the ones I host on my LAN that aren't listening on IPv6.  However, the name lookup for the directories will be returning an IPv4 address, so that's what will be used for requesting the server lists.

**Reply by softins** [reply]
Here's what I have just done:

1. On my Pi, start Jamulus without `-6`, I can see all genre lists and can connect to my private server on IPv4 but not IPv4. Restart Jamulus with `-6`. I can still see all genre lists, and can connect to my private server on both IPv4 and IPv6.
2. On my Windows machine, repeat the above tests, and find the same results.
3. On my Macbook running Mojave, repeat the above (`/Applications/Jamulus.app/Contents/MacOS/Jamulus`) with the same results. Run `/Applications/Jamulus.app/Contents/MacOS/Jamulus -6`, and again, I can still see all the lists, and can connect to my private server using both IPv4 and IPv6. So on my Mac it works fine.

Note that on my Mac I have the Firewall switched off (System Preferences > Security & Privacy > Firewall).

@rdica We will need more specific information (OS, Jamulus version, etc.), to try and understand what's happening for you. But unless it's an incompatibility with a more recent MacOS, I don't think it's a Jamulus bug.

**Comment by rdica** [comment]
```
uname -a
Darwin C02WX4BQJG5M.dhcp.ldev 19.6.0 Darwin Kernel Version 19.6.0: Sun Nov 14 19:58:51 PST 2021; root:xnu-6153.141.50~1/RELEASE_X86_64 x86_64
```
MacOS Catalina 10.15.7 fully updated.

I disabled the firewall, but that had same results. Running tcpdump on the interface: `sudo tcpdump -v -i en0 udp port 22124` I get the following from Any Genre 1:
```
tcpdump: listening on en0, link-type EN10MB (Ethernet), capture size 262144 bytes
13:18:12.694124 IP (tos 0x0, ttl 64, id 35473, offset 0, flags [none], proto UDP (17), length 37)
    c02wx4bqjg5m.dhcp.ldev.22141 > sunrise.green-lodges.ddns.net.22124: UDP, length 9
13:18:12.848155 IP (tos 0x0, ttl 42, id 52128, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2362 > 1472
13:18:12.850001 IP (tos 0x0, ttl 42, id 52129, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 4484 > 1472
13:18:14.794641 IP (tos 0x0, ttl 64, id 59406, offset 0, flags [none], proto UDP (17), length 37)
    c02wx4bqjg5m.dhcp.ldev.22141 > sunrise.green-lodges.ddns.net.22124: UDP, length 9
13:18:14.940397 IP (tos 0x0, ttl 42, id 52148, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2362 > 1472
13:18:14.942641 IP (tos 0x0, ttl 42, id 52149, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 4484 > 1472
13:18:16.840737 IP (tos 0x0, ttl 64, id 52179, offset 0, flags [none], proto UDP (17), length 37)
    c02wx4bqjg5m.dhcp.ldev.22141 > sunrise.green-lodges.ddns.net.22124: UDP, length 9
13:18:16.992834 IP (tos 0x0, ttl 42, id 52377, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2362 > 1472
13:18:16.995104 IP (tos 0x0, ttl 42, id 52378, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 4484 > 1472
13:18:18.841521 IP (tos 0x0, ttl 64, id 63948, offset 0, flags [none], proto UDP (17), length 37)
    c02wx4bqjg5m.dhcp.ldev.22141 > sunrise.green-lodges.ddns.net.22124: UDP, length 9
13:18:18.986267 IP (tos 0x0, ttl 42, id 52727, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2362 > 1472
13:18:18.989572 IP (tos 0x0, ttl 42, id 52728, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 4484 > 1472
13:18:20.840929 IP (tos 0x0, ttl 64, id 59802, offset 0, flags [none], proto UDP (17), length 37)
    c02wx4bqjg5m.dhcp.ldev.22141 > sunrise.green-lodges.ddns.net.22124: UDP, length 9
13:18:21.092735 IP (tos 0x0, ttl 42, id 52858, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2362 > 1472
13:18:21.094246 IP (tos 0x0, ttl 42, id 52859, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 4484 > 1472
13:18:22.840801 IP (tos 0x0, ttl 64, id 27859, offset 0, flags [none], proto UDP (17), length 37)
    c02wx4bqjg5m.dhcp.ldev.22141 > sunrise.green-lodges.ddns.net.22124: UDP, length 9
13:18:23.127426 IP (tos 0x0, ttl 42, id 52950, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2362 > 1472
13:18:23.130054 IP (tos 0x0, ttl 42, id 52951, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 4484 > 1472
```
This continues to scroll until I close the Connection dialog box.
There is no traffic at all when I select any other genre.

I tried a direct connect to an IPv4 IP address of my server (daw2.diix.org), I see the connect attempt in that servers log, but it appears the connection cannot be completed. 

Rerunning tcpdump with `-X` shows this while trying to view server list on Any Genre 1:
```
13:32:58.281725 IP (tos 0x0, ttl 42, id 28394, offset 0, flags [+], proto UDP (17), length 1500)
    sunrise.green-lodges.ddns.net.22124 > c02wx4bqjg5m.dhcp.ldev.22141: UDP, bad length 2363 > 1472
	0x0000:  4500 05dc 6eea 2000 2a11 1f3a 3e71 cc68  E...n...*..:>q.h
	0x0010:  c0a8 116b 566c 567d 0943 d9cb 0000 fa03  ...kVlV}.C......
	0x0020:  0032 0900 0000 0000 0010 4469 7265 6374  .2........Direct
	0x0030:  6f72 7920 5365 7276 6572 0e54 7abc 6c56  ory.Server.Tz.lV
	0x0040:  0b74 6865 6c6f 776b 6963 6b73 7b6a 8853  .thelowkicks{j.S
	0x0050:  6c56 0f46 6f72 7465 5072 656e 6573 7469  lV.FortePrenesti
	0x0060:  6e6f c8cd 40d9 6c56 0c45 6368 6f6d 7573  no..@.lV.Echomus
	0x0070:  6573 2052 4dd1 7ea2 b96c 5613 4563 686f  es.RM.~..lV.Echo
	0x0080:  6d75 7365 7320 4a65 7275 7361 6c65 6de4  muses.Jerusalem.
	0x0090:  3644 bc6c 560e 796f 7572 5365 7276 6572  6D.lV.yourServer
	0x00a0:  4e61 6d65 e37c b223 6c56 0e4d 4954 415f  Name.|.#lV.MITA_
	0x00b0:  4a61 6d53 6572 7665 7217 ce71 3e6c 560a  JamServer..q>lV.
	0x00c0:  4269 6572 696d 7075 6c73 65fa a2b9 6c56  Bierimpulse...lV
	0x00d0:  0e44 656e 6e69 7350 6c61 7952 6f6f 6d78  .DennisPlayRoomx
	0x00e0:  c4dd 236c 5607 6f6e 6534 616c 6c89 c502  ..#lV.one4all...
	0x00f0:  516c 560e 6a61 6d2e 6861 6c6d 616e 2e6e  QlV.jam.halman.n
	0x0100:  6574 a6d7 e3d4 6c56 0843 686f 7258 616e  et....lV.ChorXan
	0x0110:  6766 ce71 3e8c 590d 4f6c 6420 4469 7265  gf.q>.Y.Old.Dire
	0x0120:  6374 6f72 79c3 6b53 2d6c 560e 796f 7572  ctory.kS-lV.your
	0x0130:  5365 7276 6572 4e61 6d65 74cc 6a57 6c56  ServerNamet.jWlV
	0x0140:  124a 6f67 6954 6563 2050 412d 5465 6368  .JogiTec.PA-Tech
	0x0150:  6e69 6bbe 6d45 366c 5607 4361 7473 4157  nik.mE6lV.CatsAW
	0x0160:  5317 fdf4 506c 5607 4a61 7272 6f62 61b8  S...PlV.Jarroba.
	0x0170:  8e05 556a 5607 4350 2046 5245 451b 4af6  ..UjV.CP.FREE.J.
	0x0180:  366c 560a 4365 6f6c 4269 6e6e 2032 3687  6lV.CeolBinn.26.
	0x0190:  ec57 acfb 0841 7175 616c 756e 67b1 5841  .W...Aqualung.XA
	0x01a0:  526c 560f 5b6e 616d 655d 4a61 6d4d 7573  RlV.[name]JamMus
	0x01b0:  4e69 6264 0850 036c 5608 6272 6f62 6265  Nibd.P.lV.brobbe
	0x01c0:  7279 dff9 5e17 6c56 0e79 6f75 7253 6572  ry..^.lV.yourSer
	0x01d0:  7665 724e 616d 65e0 dbdd 596c 560a 437a  verName...YlV.Cz
	0x01e0:  6563 6869 6132 3632 6f2e a0d9 6c56 1066  echia262o...lV.f
	0x01f0:  6565 656c 732e 4a41 4d2e 776f 726c 640e  eeels.JAM.world.
	0x0200:  c9a4 d56c 560d 5265 706c 6f6b 616c 656e  ...lV.Replokalen
	0x0210:  4253 4475 ad9e 126c 5613 5433 4a61 6d20  BSDu...lV.T3Jam.
	0x0220:  4950 7636 2f70 616e 6465 6c61 7918 9384  IPv6/pandelay...
	0x0230:  5b6c 5614 4220 666c 6174 2033 2e38 2e31  [lV.B.flat.3.8.1
	0x0240:  2c34 3020 7573 7273 4514 8412 6c56 1158  ,40.usrsE...lV.X
	0x0250:  6176 734a 616d 756c 7573 5365 7276 6572  avsJamulusServer
	0x0260:  d779 2372 6e56 0873 756c 756d 616a 61f0  .y#rnV.sulumaja.
	0x0270:  05f7 c76c 5609 4469 6520 496e 7365 6c44  ...lV.Die.InselD
	0x0280:  424b 336c 560e 796f 7572 5365 7276 6572  BK3lV.yourServer
	0x0290:  4e61 6d65 7ca4 335c 6c56 0946 494c 5445  Name|.3\lV.FILTE
	0x02a0:  524c 4f53 5da0 68ac 6c56 1265 7a6a 616d  RLOS].h.lV.ezjam
	0x02b0:  3a20 4e61 6e6f 6465 2031 2047 424d b8d7  :.Nanode.1.GBM..
	0x02c0:  236c 5613 484b 4b65 6c76 696e 2d4a 616d  #lV.HKKelvin-Jam
	0x02d0:  756c 7573 2d30 3197 dc39 b06c 5608 4765  ulus-01..9.lV.Ge
	0x02e0:  6f73 3234 2f37 8b4d 0803 6c56 0e48 7567  os24/7.M..lV.Hug
	0x02f0:  6865 6e64 656e 4661 7274 73d8 5553 8c6c  hendenFarts.US.l
	0x0300:  560e 4a61 6d4f 7261 636c 6543 6c6f 7564  V.JamOracleCloud
	0x0310:  4278 6d6d 6c56 0b4a 616d 2061 7420 416b  BxmmlV.Jam.at.Ak
	0x0320:  2773 1be6 8103 6c56 0fe2 809c 4243 4d5f  's....lV....BCM_
	0x0330:  4a61 6d75 6c75 7332 b69f 5536 6c56 1142  Jamulus2..U6lV.B
	0x0340:  6c75 6d4a 616d 756c 7573 5365 7276 6572  lumJamulusServer
	0x0350:  9d52 9d05 6c56 144a 616d 756c 7573 5365  .R..lV.JamulusSe
	0x0360:  7276 6572 2e63 6f6d 3030 34a4 a9e5 366c  rver.com004...6l
	0x0370:  560c 5361 6c74 7967 726f 776c 6572 0de3  V.Saltygrowler..
	0x0380:  5a9d 6c56 076a 616d 3446 756e 8099 fcc3  Z.lV.jam4Fun....
	0x0390:  6c56 0a50 6c75 7353 6572 7665 7213 326f  lV.PlusServer.2o
	0x03a0:  1f6c 560b 5374 6f72 6d53 6572 7665 7207  .lV.StormServer.
	0x03b0:  75ca 906c 5614 4d72 2e4c 6577 6973 4a61  u..lV.Mr.LewisJa
	0x03c0:  6d75 6c75 7353 6572 7665 bd2d fc22 6c56  mulusServe.-."lV
	0x03d0:  0b43 4d4a 616d 5365 7276 6572 a94b 9336  .CMJamServer.K.6
	0x03e0:  6c56 0954 534f 2d43 6f6d 626f 63ab b836  lV.TSO-Comboc..6
	0x03f0:  6c56 0a41 5055 4a61 6d75 6c75 7329 11a1  lV.APUJamulus)..
	0x0400:  346c 5608 4a45 5342 616e 6a6f a81f 7032  4lV.JESBanjo..p2
	0x0410:  6c56 0646 616d 4a61 6dfe d5b2 236c 5607  lV.FamJam...#lV.
	0x0420:  5068 696c 4a61 6d68 9eb9 926c 5606 416e  PhilJamh...lV.An
	0x0430:  6465 7273 03b2 5122 6c56 104a 656e 735f  ders..Q"lV.Jens_
	0x0440:  4661 7374 2053 6572 7665 72bd 6555 036c  Fast.Server.eU.l
	0x0450:  5606 4a61 6d50 6164 69e9 6e8f 6c56 1450  V.JamPadi.n.lV.P
	0x0460:  6163 6966 6963 426c 7565 6772 6173 7326  acificBluegrass&
	0x0470:  4f6c 647e c768 d06c 5606 5343 204a 616d  Old~.h.lV.SC.Jam
	0x0480:  a4d1 ffd5 6c56 0f4e 6577 486f 7269 7a6f  ....lV.NewHorizo
	0x0490:  6e73 4261 6e64 90be 092d 6c56 0e4d 6564  nsBand...-lV.Med
	0x04a0:  696f 416c 6561 746f 7269 6f60 73d4 8d6c  ioAleatorio`s..l
	0x04b0:  5612 6a61 6d2e 6565 6373 2e75 6d69 6368  V.jam.eecs.umich
	0x04c0:  2e65 6475 15ce dd34 7656 0753 4f46 544a  .edu...4vV.SOFTJ
	0x04d0:  414d bab3 e003 6c56 124a 6170 616e 6573  AM....lV.Japanes
	0x04e0:  654b 6172 616f 6b65 5553 417c f7ae 9e6c  eKaraokeUSA|...l
	0x04f0:  560b 546f 6d73 6569 694a 616d 73e2 195f  V.TomseiiJams.._
	0x0500:  966c 5626 f09f 8f98 20e0 b8ab e0b9 89e0  .lV&............
	0x0510:  b8ad e0b8 87e0 b88b e0b9 89e0 b8ad e0b8  ................
	0x0520:  a1e0 b8af 204c 6f62 6279 b24c 2705 6c56  .....Lobby.L'.lV
	0x0530:  0744 504b 5072 6f64 563d 37a2 6c56 0e54  .DPKProdV=7.lV.T
	0x0540:  6865 2043 6972 636c 6520 5069 7424 cce3  he.Circle.Pit$..
	0x0550:  d46c 560b 5468 654c 6573 7369 6e67 7371  .lV.TheLessingsq
	0x0560:  2937 536c 5600 d2ce 0a42 6c56 134d 6964  )7SlV....BlV.Mid
	0x0570:  6e69 6768 7420 526f 6465 6f20 4261 6e64  night.Rodeo.Band
	0x0580:  5af3 96cb 6c56 2de0 b8ab e0b9 89e0 b8ad  Z...lV-.........
	0x0590:  e0b8 87e0 b88b e0b9 89e0 b8ad e0b8 a1e0  ................
	0x05a0:  b8af 205b e0b8 aae0 b8b2 e0b8 82e0 b8b2  ...[............
	0x05b0:  206f 315d 7869 b723 6c56 0852 4353 2d4a  .o1]xi.#lV.RCS-J
	0x05c0:  616d 31d7 cbfc 673a fe04 4d61 7273 2b9e  am1...g:..Mars+.
	0x05d0:  dc23 6c56 15f0 9f8e 8459 616f            .#lV.....Yao
```
which drew my attention to the packet size, and `bad length` indications in each packet of the initial dump.

Looks to me like MTU issues?
This Mac is on WIFI, could that be the cause?

If there's any additional info I can provide, please let me know.
Meantime I'm going to see if I can get this Mac on a wired connection and retest.

**Reply by pljones** [reply]
Can you run that with numeric addresses rather than names, so it's clearer whether IPv4 or IPv6 is in use, please.  The bad length looks... interesting, indeed.

**Reply by rdica** [reply]
It's IPv4, as indicated by the protocol version after the timestamp (`13:18:23.130054 IP (tos 0x0,...`). Were they IPv6, the protocol version would show as `IP6` (`14:15:08.743984 IP6 (class 0xc0,...`).

**Reply by pljones** [reply]
Ah thanks, missed that bit.

**Reply by softins** [reply]
Could you do a capture to a file using `tcpdump -w file -s 0` (plus any other options required)? Then we could examine it in Wireshark. As well as Jamulus traffic, please include DNS traffic too.

**Reply by softins** [reply]
If you use Wireshark, you might find https://github.com/softins/jamulus-wireshark of interest too.

**Reply by rdica** [reply]
@softins  thanks, I already installed that when I was testing some non-related things. I will work on testing with hardwired connection, and also do captures on both WIFI and hardwired.

**Reply by rdica** [reply]
I didn't change anything, but I found I had another running jamulus process during all my testing, I killed it, restarted with `-6` option and I now see list of servers in Connection dialog. I can direct connect via IPv4 and IPv6 as well.

I also found, much to my chagrine, that my ethernet adapter for Macbook was for an older version, it uses thunderbolt connector and my current macbook doesnt have that. _sigh_.

I'm closing this as user error, sorry for wasting everyones time that looked at this.

**Reply by softins** [reply]
@rdica thanks for the update. No problem, and it's still useful to know that this can occur.

I have a hypothesis, which I have not tested: I suspect that the operating system may have allowed the second instance, when opening an `AF_INET6` socket, to bind to the same UDP port number that the first instance had bound to on an `AF_INET` socket. Then when the second instance sent out IPv4 traffic (using the valid mapping of IPv4 address within an IPv6 address), the replies from the remote host got routed to the first instance instead of the second.

**Comment by frank-valentin** [comment]
Hi, there! 
I'm quite lost. I'm trying to find out which version of Jamulus works with MacOS Catalina 10.15.7
The current version I have downloaded tells me I need MacOS 11 or later
The jamulus user knowledge database first pointed me to sourceforge, then here. Sorry if this is the wrong place/thread. Can anybody please help?
thanks!

**Reply by chrisrimple** [reply]
You need the "legacy version" for MacOS.  Visit https://github.com/jamulussoftware/jamulus/releases/tag/r3_11_0 and scroll down to the list of Assets.  You'll see "[jamulus_3.11.0_mac_legacy.dmg](https://github.com/jamulussoftware/jamulus/releases/download/r3_11_0/jamulus_3.11.0_mac_legacy.dmg)", which is the build you want.  Download the DMG file, then double-click to run the installer.

**Reply by frank-valentin** [reply]
Thanks ever so much!!!!

**Reply by frank-valentin** [reply]
Damn! When I try to open it I get the following message:

“Jamulus” can’t be opened because Apple cannot check it for malicious software.
This software needs to be updated. Contact the developer for more information.

What do I do?

**Reply by softins** [reply]
> Damn! When I try to open it I get the following message:
> 
> “Jamulus” can’t be opened because Apple cannot check it for malicious software. This software needs to be updated. Contact the developer for more information.
> 
> What do I do?

You need to go to Finder, into Applications, and do Open on the Jamulus application. It will most likely give you that message, with no Open button, but a "Show in Finder" button. Click on "Show in Finder", and it will display the Applications folder again. This time, do Open again, and it should offer you an additional button for "Open". Click on "Open" and it will start the application. It will also remember that now, and allow you to open it normally each time without the warning.

You will need to repeat this procedure each time you install a new version.

**Reply by frank-valentin** [reply]
Hurrah!!! It is working now
Thank you softins and chrisrimple for your kind and swift help
